src/couch/src/couch_db.erl

```diff
@@ -945,13 +945,14 @@ prep_and_validate_replicated_updates(Db, [Bucket|RestBuckets], [OldInfo|RestOldI
     {ok, #full_doc_info{rev_tree=OldTree}} ->
         OldLeafs = couch_key_tree:get_all_leafs_full(OldTree),
         OldLeafsLU = [{Start, RevId} || {Start, [{RevId, _}|_]} <- OldLeafs],
+        NewPaths = lists:map(fun couch_doc:to_path/1, Bucket),
+        SortedPaths = lists:sort(NewPaths),
```

Comment:
Its not because it is necessary for correctness. Granted it's a super edge case scenario but the underlying issue is exactly the same as from COUCHDB-1163 [1].

In a nutshell, when merging multiple paths at once we could have this set of paths (written in short hand):

    {1, [A, B, C]}
    {3, [C, D, E]}
    {5, [E, F, G]}

If we merged them without sorting we could end up with something like:

    {5, [E, F, G]}
    {1, [A, B, C]}
    {3, [C, D, E]}

Which when merged gives us these two branches:

    {1, [A, B, C, D, E]}
    {5, [E, F, G]}

The legacy implementation of merge/3 calls stem for us which fixes the issue forcefully by tearing apart all paths, sorting them, and then merges them back together into a final tree. If you read through the commits where I tweaked the unit tests those were broken due to us not even realizing how much that stem operation was "fixing" the bad merges.

Hopefully this discussion doesn't trigger any night terrors of bugs of yore.

https://issues.apache.org/jira/browse/COUCHDB-1163?focusedCommentId=13035172&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13035172

Reply:
Oooh, I missed that you switch from `merge/3` to `merge/2` here. Thanks.

Reply:
Ah, sorry! I should probably call that out better as its been missed by most everyone reviewing this. Though I'd really like to figure out a way to move all of the sorting/stem-or-not logic into the key tree so am not pushing this too hard while I mull over it.

Reply:
Totally agree. I'm also not a fan of the stem vs full_stem differences as well. I'm contemplating merging the two and carrying a set of revision keys through the optimized O(N) version and then bailing and calling "full_stem" if there's a duplicate key detected. Though am a bit distracted by other deadlines at the moment so not devoting full thought to this.